### PR TITLE
added support for using (non-branch) refs

### PR DESCRIPTION
**Pull request by bpeabody**
Previously could only talk about branches. This work can land on `master` now as it does not require or imply rebranding.

### CODE REVIEW

File: node/lib/util/repo_ast.js [hunk 1]
```diff
@@ -224,13 +224,15 @@ class AST {
...
+     *   specified, exists in the `commits` map.
+     * - no names in `refs` indicate branches: local (starting with 'heads/' or
+     *   remote, (starting with 'remotes/')
      * - `currentBranchName`, if specified, exists in `branches`
      * - each commit in `commits` is *reachable* from at least one reference in
-     *   `branch`, by `head`, or by a remote branch that is, it is pointed to
-     *   by a reference or is a direct or indirect ancestor of a commit that
-     *   is.
+     *   `branch`, by `head`, by a reference in `refs`, or by a remote branch
+     *   that is, it is pointed to by a reference or is a direct or indirect
```
Comment (novalis) [inline node/lib/util/repo_ast.js]: I can't parse this.

File: node/lib/util/repo_ast.js [hunk 2]
```diff
@@ -244,6 +246,8 @@ class AST {
      * @param {Object}      args
      * @param {Object}      [args.commits]
      * @param {Object}      [args.branches]
+     * @param {Object}      [args.refs]
+     * @param {Object}      [args.refs]
```
Comment (novalis) [inline node/lib/util/repo_ast.js]: duplicate line

File: node/lib/util/repo_ast.js [hunk 3]
```diff
@@ -335,6 +339,21 @@ in commit ${id}.`);
...
+        this.d_refs = {};
+        const refs = ("refs" in args) ? args.refs : {};
+        for (let name in refs) {
+            const ref = refs[name];
+            assert.isString(ref, name);
+            assert(!name.startsWith("heads/"),
+                   `ref ${name} indicates a branch`);
+            assert(!name.startsWith("remotes/"),
+                   `ref ${name} indicates a remote branch`);
+            checkAndTraverse(ref, `ref ${ref}`);
```
Comment (novalis) [inline node/lib/util/repo_ast.js]: should this be `ref ${name}` ?

File: node/lib/util/repo_ast_util.js [hunk 4]
```diff
@@ -300,7 +300,32 @@ ${colorExp(expectedBranch)}`
                 missingExpectedBranch,
                 compareBranches);
 
+    // Then check refs
+
+    function missingActualRef(ref) {
+        result.push(`missing ref ${colorBad(ref)}`);
+    }
+    function missingExpectedRef(ref) {
+        result.push(`unexpected ref ${colorBad(ref)}`);
```
Comment (novalis) [inline node/lib/util/repo_ast_util.js]: this function's name and message do not match.

File: node/lib/util/shorthand_parser_util.js [hunk 5]
```diff
@@ -61,6 +61,7 @@ const RepoASTUtil  = require("../util/repo_ast_util");
  * nothing        =
  * commit         = <alphanumeric>+
  * branch         = 'B'<name>'='<commit>|<nothing>     nothing deletes branch
+ * ref            = 'F'<name>'='<commit>|<nothing>     nothing deletes branch
```
Comment (novalis) [inline node/lib/util/shorthand_parser_util.js]: "nothing deletes ref"

File: node/lib/util/shorthand_parser_util.js [hunk 6]
```diff
@@ -296,6 +298,18 @@ function prepareASTArguments(baseAST, rawRepo) {
         }
     });
 
+    // Copy in ref  overrides, deleting where `null` was specified.
```
Comment (novalis) [inline node/lib/util/shorthand_parser_util.js]: nit: there's an extra space after ref

File: node/lib/util/shorthand_parser_util.js [hunk 7]
```diff
@@ -456,36 +471,40 @@ function parseOverrides(shorthand, begin, end, delimiter) {
     }
 
     /**
-     * Parse the branch override beginning at the specified `begin` and ending
-     * at the specified `end`.
+     * Return a parser for the specified type of `reference`; copy the results
+     * into the specifeid `dest` object.
```
Comment (novalis) [inline node/lib/util/shorthand_parser_util.js]: nit: spelling on specifeid.

File: node/lib/util/repo_ast.js (hunk 1 above)
Comment (bpeabody) [inline node/lib/util/repo_ast.js]: reworking

File: node/lib/util/repo_ast.js (hunk 2 above)
Comment (bpeabody) [inline node/lib/util/repo_ast.js]: ack

**Comment by bpeabody** [comment]
Should be good-to-go. One requested change about function message and name being mismatched I think was understandable but not right; the function name described intent but the message was appropriate for displaying to user.

**Pull request merged**